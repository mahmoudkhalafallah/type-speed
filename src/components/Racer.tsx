import React, { useEffect, useState } from 'react'

import styled from 'styled-components'
import Quote from './Quote'
import { QUOTE_URL } from '../constants'

const UserEntry = styled.input`
width: 100%;
padding: 14px;
margin-top: 15px;
font-size: 20px;
display: block;
border: 1px solid #ffb800;
`

const NoInteractions = styled.div`
background: rgba(238, 238, 238, 0.5);
width: 100%;
height: 100%;
position: absolute;
`
const GameContainer = styled.section`
position: relative
`

interface Props {
    gameOver: boolean;
    endGame: React.Dispatch<React.SetStateAction<boolean>>;
    setWpm: React.Dispatch<React.SetStateAction<number>>;
    setCp: React.Dispatch<React.SetStateAction<number>>;
}

const Racer: React.FC<Props> = ({ gameOver, endGame, setWpm, setCp }) => {
    const [quoteText, setQuoteText] = useState([''])
    const [typedText, setTypedText] = useState('')
    const [wordIndex, setWordIndex] = useState(0)
    const [typedTextValiditiy, setTypedTextValiditiy] = useState(false)
    const [secondsElapsed, setSecondsElapsed] = useState(0)

    const handleChange = (e: any) => {
        const val = e.target.value
        setTypedTextValiditiy(val.length > 0 && quoteText[wordIndex].startsWith(val))

        if (e.nativeEvent.data === ' ' && quoteText[wordIndex] === val.slice(0, -1)) {
            setTypedText('')
            if (wordIndex === quoteText.length - 1) {
                endGame(true)
            }
            setWordIndex(c => c + 1)
        } else {

            setTypedText(val)
        }
    }

    useEffect(() => {
        fetch(QUOTE_URL).then(res => res.text()).then((data: string) => {
            data = data.replace(/\s{2,}/g, ' ') // remove extra spaces
            setQuoteText(data.split(" "))
        })
    }, [])

    useEffect(() => {
        const timerHandle = setInterval(() => {
            if (gameOver) {
                clearInterval(timerHandle)
                const calculateCp = (quoteText.slice(0, wordIndex).join(' ').length / quoteText.join(' ').length) * 100
                setCp(parseInt(calculateCp.toFixed()))
                return
            }
            setSecondsElapsed(s => s + 1)
            const typedChars = quoteText.slice(0, wordIndex).join(' ').length
            if (secondsElapsed > 0) {
                setWpm((typedChars / 5) / (secondsElapsed / 60))
            }
        }, 1000)
        return () => {
            clearInterval(timerHandle)
        }
    }, [gameOver, quoteText, secondsElapsed, setCp, setWpm, wordIndex])

    return (
        <GameContainer>
            {gameOver && <NoInteractions />}
            {(quoteText.length > 1) &&
                <>
                    <Quote data={quoteText} index={wordIndex} length={typedText.length} isValid={typedTextValiditiy} />
                    <UserEntry value={typedText} onChange={handleChange} readOnly={gameOver} />
                </>
            }
        </GameContainer>
    )
}

export default Racer