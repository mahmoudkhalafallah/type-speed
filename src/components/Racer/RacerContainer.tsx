// eslint-disable-next-line no-unused-vars
import React, { ChangeEvent } from 'react'
import styled from 'styled-components'
import Quote from './Quote'

const UserEntry = styled.input`
width: 100%;
padding: 14px;
margin-top: 15px;
font-size: 20px;
display: block;
border: 1px solid #ffb800;
`

const NoInteractions = styled.div`
background: rgba(238, 238, 238, 0.5);
width: 100%;
height: 100%;
position: absolute;
`
const StyledRacerContainer = styled.section`
position: relative;
`

interface Props {
    gameOver: boolean;
    quoteText: { text: string[], count: number };
    wordIndex: number;
    typedText: string;
    typedTextValiditiy: boolean;
    handleChange: (e: ChangeEvent) => void
}

const RacerContainer: React.FC<Props> = ({ gameOver, quoteText, wordIndex, typedText, typedTextValiditiy, handleChange }) =>
    <StyledRacerContainer>
        {gameOver && <NoInteractions />}
        {(quoteText.count > 0) &&
            <>
                <Quote
                    data={quoteText.text}
                    index={wordIndex}
                    length={typedText.length}
                    isValid={typedTextValiditiy}
                />
                <UserEntry
                    value={typedText}
                    onChange={handleChange}
                    readOnly={gameOver}
                />
            </>
        }
    </StyledRacerContainer>
export default RacerContainer