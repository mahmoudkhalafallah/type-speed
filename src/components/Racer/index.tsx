import React, { useEffect, useState } from 'react'
import { QUOTE_URL } from '../../constants'
import RacerContainer from './RacerContainer'

const AVERAGE_WORD_LENGTH = 5 // "word" is often standardized to be five characters or keystrokes long in English
const SECONDS_IN_MINUTE = 60 // number of seconds in a minute

interface Props {
    gameOver: boolean;
    endGame: React.Dispatch<React.SetStateAction<boolean>>;
    setWpm: React.Dispatch<React.SetStateAction<number>>;
    setCp: React.Dispatch<React.SetStateAction<number>>;
}

const Racer: React.FC<Props> = ({ gameOver, endGame, setWpm, setCp }) => {
    const [quoteText, setQuoteText] = useState({ text: [''], count: 0 })
    const [typedText, setTypedText] = useState('')
    const [wordIndex, setWordIndex] = useState(0)
    const [typedTextValiditiy, setTypedTextValiditiy] = useState(false)
    const [secondsElapsed, setSecondsElapsed] = useState(0)

    const handleChange = (e: any) => {
        const val = e.target.value
        setTypedTextValiditiy(val.length > 0 && quoteText.text[wordIndex].startsWith(val))

        if (e.nativeEvent.data === ' ' && quoteText.text[wordIndex] === val.slice(0, -1)) { // if user enters a space and the typed word matches the quote word
            setTypedText('') // clear the current typed text
            if (wordIndex === quoteText.text.length - 1) { // if the accepted word is the last word in the quote end the game
                endGame(true)
                return
            }
            setWordIndex(c => c + 1) // and update the current word index
        } else { // otherwise just update the typed text
            setTypedText(val)
        }
    }

    useEffect(() => {
        fetch(QUOTE_URL).then(res => res.text()).then((data: string) => {
            data = data.replace(/\s{2,}/g, ' ') // remove extra spaces
            setQuoteText({ text: data.split(" "), count: data.length })
        })
    }, [])

    useEffect(() => {
        const timerHandle = setInterval(() => {
            const typedCharsCount = quoteText.text.slice(0, wordIndex).join(' ').length

            // Calculate Words Per Minute rate
            if (secondsElapsed > 0 && typedCharsCount > 0) {
                const wordsCount = (typedCharsCount / AVERAGE_WORD_LENGTH)
                const minutesCount = (secondsElapsed / SECONDS_IN_MINUTE)
                const wpm = (wordsCount / minutesCount).toFixed()

                setWpm(parseInt(wpm))
            }

            if (gameOver) {
                clearInterval(timerHandle)
                // Calculate Completion Percentage
                const calculateCp = (typedCharsCount / quoteText.count) * 100
                setCp(parseInt(calculateCp.toFixed()))

                return
            }
            setSecondsElapsed(e => e + 1)

        }, 1000)
        return () => {
            clearInterval(timerHandle)
        }
    }, [gameOver, quoteText, secondsElapsed, setCp, setWpm, wordIndex])

    return (
        <RacerContainer
            gameOver={gameOver}
            handleChange={handleChange}
            quoteText={quoteText}
            typedText={typedText}
            typedTextValiditiy={typedTextValiditiy}
            wordIndex={wordIndex}
        />
    )
}

export default Racer