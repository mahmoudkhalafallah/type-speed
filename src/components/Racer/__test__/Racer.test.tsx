// eslint-disable-next-line no-unused-vars
import React, { Dispatch, SetStateAction } from 'react'
import Racer from '..'
import { render } from '@testing-library/react'

const endGame: Dispatch<SetStateAction<boolean>> = jest.fn((val) => val)
const setWpm: Dispatch<SetStateAction<number>> = jest.fn((val) => val)
const setCp: Dispatch<SetStateAction<number>> = jest.fn((val) => val)

test('Racer rendered correctly', () => {
    const { baseElement } = render(
        <Racer gameOver={false} endGame={endGame} setCp={setCp} setWpm={setWpm} />,
    )

    expect(baseElement).toMatchSnapshot()
})
