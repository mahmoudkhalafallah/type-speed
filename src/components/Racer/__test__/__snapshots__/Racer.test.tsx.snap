// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Racer rendered correctly 1`] = `
<body>
  <div>
    <section
      class="sc-bZQynM dCMySu"
    />
  </div>
</body>
`;
